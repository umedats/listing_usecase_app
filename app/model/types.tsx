import { Subscription as SDKSubscription } from "@/app/sdk/notifications";

/**
 * The Issue returned from the Listing API.
 */
export interface Issue {
  code: string;
  message: string;
  severity: string;
  attributeNames?: string[];
}

/**
 * The result of the Listings Items API submission.
 */
export interface ListingSubmissionResult {
  sku: string;
  status: string;
  submissionId: string;
  issues?: Issue[];
}

/**
 * Relevant information from Listings GET API call.
 */
export interface Listing {
  attributes: object;
  issues?: Issue[];
  productType?: string;
}

/**
 * A type which represents a patch for the contribution.
 */
export interface Patch {
  op: string;
  path: string;
  value: any;
}

/**
 * A type for the alert dialog state.
 */
export interface AlertDialogState {
  showAlert: boolean;
  alertTitle: string;
  alertContent: string;
}

/**
 * A type which captures the SP API request and response.
 */
export interface SPAPIRequestResponse {
  apiName: string;
  apiDocumentationLink: string;
  request: object;
  response: object;
}

/**
 * A type which capture the information about a feed.
 */
export interface Feed {
  feedId: string;
  feedType: string;
  createdTime: string;
  processingStatus: string;
  resultFeedDocumentId: string;
}

/**
 * A type which capture the information about a Subscription.
 */
export interface Subscription {
  notificationType: string;
  eventBusName?: string;
  eventBusFilterRule?: string;
  sqsTargetId?: string;
  sqsQueueName?: string;
  subscriptionId: string;
}

/**
 * A type which captures the raw SDK subscription along with the
 * notification type.
 */
export interface RawSubscription {
  notificationType: string;
  subscription: SDKSubscription;
}
