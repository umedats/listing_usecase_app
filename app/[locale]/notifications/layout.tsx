"use client";
import React from "react";
import DualPageLayout from "@/app/components/dual-page-layout";

/**
 * Parallel Route layout for the Notifications.
 * @param props input props.
 * @constructor
 */
export default function Layout(props: {
  createSubscription: React.ReactNode;
  viewSubscriptions: React.ReactNode;
}) {
  return (
    <DualPageLayout
      leftContent={props.createSubscription}
      rightContent={props.viewSubscriptions}
    />
  );
}
