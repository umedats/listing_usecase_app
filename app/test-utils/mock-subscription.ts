import { Subscription } from "@/app/model/types";
import {
  CREATE_SUBSCRIPTION_API_DOC_LINK,
  CREATE_SUBSCRIPTION_API_NAME,
} from "@/app/constants/global";

export const MOCK_SUBSCRIPTION_1: Subscription = {
  subscriptionId: "f8c3de3d-1fea-4d7c-a8b0-29f63c4c3454",
  notificationType: "BRANDED_ITEM_CONTENT_CHANGE",
  eventBusName: "TestEventBus",
  eventBusFilterRule: "SellingPartnerNotificationsAPIRule",
  sqsQueueName: "BrandedItemContentChange",
};

export const MOCK_SUBSCRIPTION_2: Subscription = {
  subscriptionId: "510f082f-a853-4847-acf2-475641cb7b06",
  notificationType: "PRODUCT_TYPE_DEFINITIONS_CHANGE",
  eventBusName: "TestEventBus1",
  eventBusFilterRule: "SellingPartnerNotificationsAPIRule",
  sqsQueueName: "ProductTypeDefinitionsChange",
};

export const MOCK_SUBSCRIPTION_3: Subscription = {
  subscriptionId: "510f082f-a853-4847-acf2-475641cb7b06",
  notificationType: "PRODUCT_TYPE_DEFINITIONS_CHANGE",
  eventBusName: "TestEventBus1",
  eventBusFilterRule: "SellingPartnerNotificationsAPIRule",
  sqsTargetId: "ProductTypeDefinitionsTarget",
  sqsQueueName: "ProductTypeDefinitionsChange",
};

export const MOCK_SUBSCRIPTION_4: Subscription = {
  subscriptionId: "510f082f-a853-4847-acf2-475641cb7b06",
  notificationType: "PRODUCT_TYPE_DEFINITIONS_CHANGE",
  eventBusName: "TestEventBus1",
  eventBusFilterRule: "SellingPartnerNotificationsAPIRule",
};

export const MOCK_CREATE_SUBSCRIPTION_RESULT_1 = {
  subscriptionId: "510f082f-a853-4847-acf2-475641cb7b06",
  payloadVersion: "1.0",
  destinationId: "testDestinationId",
};

export const MOCK_CREATE_SUBSCRIPTION_SP_API_RESPONSE = {
  apiName: CREATE_SUBSCRIPTION_API_NAME,
  apiDocumentationLink: CREATE_SUBSCRIPTION_API_DOC_LINK,
  request: {
    body: {
      payloadVersion: "1",
      destinationId: "testDestinationId",
    },
    notificationType: "LISTINGS_ITEM_STATUS_CHANGE",
  },
  response: MOCK_CREATE_SUBSCRIPTION_RESULT_1,
};
