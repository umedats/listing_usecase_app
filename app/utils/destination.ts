import { SPAPIRequestResponse } from "@/app/model/types";
import getSPAPIEndpoint, {
  buildGrantLessNotificationsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  CREATE_DESTINATION_API_DOC_LINK,
  CREATE_DESTINATION_API_NAME,
  EVENT_BRIDGE_DESTINATION_NAME,
  GET_DESTINATIONS_API_DOC_LINK,
  GET_DESTINATIONS_API_NAME,
} from "@/app/constants/global";
import {
  CreateDestinationRequest,
  Destination,
  DestinationList,
  GetDestinationsResponse,
} from "@/app/sdk/notifications";

/**
 * Helper method which invokes the GetDestinations API to fetch all the
 * event bridge destinations.
 * @param region the AWS region.
 * @param awsAccountId the AWS awsAccountId.
 * @param reqResponseCollector collects all the SP API requests and responses.
 */
export async function getEventBridgeDestinations(
  region: string,
  awsAccountId: string,
  reqResponseCollector: SPAPIRequestResponse[],
) {
  const notificationsApi = await buildGrantLessNotificationsAPIClient(
    getSPAPIEndpoint(region),
  );
  const requestResponse: SPAPIRequestResponse = {
    apiName: GET_DESTINATIONS_API_NAME,
    apiDocumentationLink: GET_DESTINATIONS_API_DOC_LINK,
    request: {},
    response: {},
  };

  try {
    const getDestinationsResponse: GetDestinationsResponse =
      GetDestinationsResponse.constructFromObject(
        await notificationsApi.getDestinations(),
        undefined,
      );
    requestResponse.response = getDestinationsResponse;
    const destinations: Destination[] = DestinationList.constructFromObject(
      getDestinationsResponse.payload,
      undefined,
    );

    return destinations?.filter(
      (destination) =>
        destination?.resource?.eventBridge?.accountId === awsAccountId &&
        destination?.resource?.eventBridge?.region === region,
    );
  } catch (error: any) {
    requestResponse.response = error;
    if (!(error?.status === 404)) {
      throw error;
    }
  } finally {
    reqResponseCollector.push(requestResponse);
  }
}

/**
 * Helper method which invokes the CreateDestination API to create new event bridge destination.
 * @param awsAccountId the AWS awsAccountId.
 * @param region the AWS region.
 * @param reqResponseCollector collects all the SP API requests and responses.
 */
export async function createEventBridgeDestination(
  awsAccountId: string,
  region: string,
  reqResponseCollector: SPAPIRequestResponse[],
) {
  const notificationsApi = await buildGrantLessNotificationsAPIClient(
    getSPAPIEndpoint(region),
  );
  const destinationResourceSpecification = {
    eventBridge: {
      region: region,
      accountId: awsAccountId,
    },
  };
  const createDestinationRequestBody =
    CreateDestinationRequest.constructFromObject(
      {
        resourceSpecification: destinationResourceSpecification,
        name: EVENT_BRIDGE_DESTINATION_NAME,
      },
      undefined,
    );
  const requestResponse = {
    apiName: CREATE_DESTINATION_API_NAME,
    apiDocumentationLink: CREATE_DESTINATION_API_DOC_LINK,
    request: {
      body: createDestinationRequestBody,
    },
    response: {},
  };

  try {
    requestResponse.response = await notificationsApi.createDestination(
      requestResponse.request.body,
    );
  } catch (error: any) {
    requestResponse.response = error;
    throw error;
  } finally {
    reqResponseCollector.push(requestResponse);
  }
  return requestResponse.response;
}
