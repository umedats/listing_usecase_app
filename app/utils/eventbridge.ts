import {
  EventBridge,
  ListRulesCommandOutput,
  ListTargetsByRuleCommandOutput,
  RemoveTargetsCommandOutput,
  ResourceAlreadyExistsException,
  ResourceNotFoundException,
  Rule,
  RuleState,
  Target,
} from "@aws-sdk/client-eventbridge";
import { serializeToJsonString } from "@/app/utils/serialization";
import { buildEventBridgeClient } from "@/app/utils/aws-clients-factory";
import {
  buildEventBusFilterRuleName,
  buildNotificationSQSQueueArn,
  buildNotificationSQSTargetName,
} from "@/app/utils/aws";
import { createQueue } from "@/app/utils/sqs";

function handleEventBridgeError(error: any) {
  console.log(serializeToJsonString(error));
  if (!(error instanceof ResourceNotFoundException)) {
    throw {
      status: 500,
    };
  }
}

/**
 * Helper method which returns the [Rule, Target] information for the given
 * eventbus.
 * @param region the aws region.
 * @param eventBusName the event bus name.
 */
export async function getRuleTargetsForEventBus(
  region: string,
  eventBusName: string,
) {
  const ruleTargets: { rule: Rule; target: Target }[] = [];
  const eventBridge = buildEventBridgeClient(region);
  try {
    let listRulesNextToken = undefined;
    do {
      const listRulesCommandOutput: ListRulesCommandOutput =
        await eventBridge.listRules({
          EventBusName: eventBusName,
          NextToken: listRulesNextToken,
        });
      listRulesNextToken = listRulesCommandOutput?.NextToken;

      const rules = listRulesCommandOutput?.Rules ?? [];
      for (const rule of rules) {
        let listTargetsByRuleNextToken = undefined;
        do {
          const listTargetsByRuleCommandOutput: ListTargetsByRuleCommandOutput =
            await eventBridge.listTargetsByRule({
              EventBusName: eventBusName,
              Rule: rule.Name,
            });
          listTargetsByRuleNextToken =
            listTargetsByRuleCommandOutput?.NextToken;
          const targets = listTargetsByRuleCommandOutput?.Targets ?? [];
          for (const target of targets) {
            ruleTargets.push({
              rule: rule,
              target: target,
            });
          }
        } while (listTargetsByRuleNextToken);
      }
    } while (listRulesNextToken);
  } catch (error) {
    handleEventBridgeError(error);
  }
  return ruleTargets;
}

async function removeTarget(
  eventBridge: EventBridge,
  eventBusName: string,
  rule: string,
  targetId: string,
) {
  try {
    const removeTargetsCommandOutput: RemoveTargetsCommandOutput =
      await eventBridge.removeTargets({
        EventBusName: eventBusName,
        Rule: rule,
        Ids: [targetId],
      });

    if (removeTargetsCommandOutput?.FailedEntryCount !== 0) {
      removeTargetsCommandOutput.FailedEntries?.forEach((failedEntry) => {
        console.log(serializeToJsonString(failedEntry));
      });
      handleEventBridgeError("removeTargets API returned 1 FailedEntryCount");
    }
  } catch (error) {
    handleEventBridgeError(error);
  }
}

async function createTarget(
  eventBridge: EventBridge,
  eventBusName: string,
  region: string,
  awsAccountId: string,
  notificationType: string,
) {
  try {
    const putTargetsResponse = await eventBridge.putTargets({
      EventBusName: eventBusName,
      Rule: buildEventBusFilterRuleName(notificationType),
      Targets: [
        {
          Arn: buildNotificationSQSQueueArn(
            region,
            awsAccountId,
            notificationType,
          ),
          Id: buildNotificationSQSTargetName(notificationType),
        },
      ],
    });

    if (putTargetsResponse?.FailedEntryCount !== 0) {
      putTargetsResponse.FailedEntries?.forEach((failedEntry) => {
        console.log(serializeToJsonString(failedEntry));
      });
      handleEventBridgeError(
        "putTargets API returned at least 1 FailedEntryCount",
      );
    }
  } catch (error: any) {
    if (!(error instanceof ResourceAlreadyExistsException)) {
      handleEventBridgeError(error);
    }
  }
}

async function deleteRule(
  eventBridge: EventBridge,
  eventBusName: string,
  rule: string,
) {
  try {
    await eventBridge.deleteRule({
      EventBusName: eventBusName,
      Name: rule,
    });
  } catch (error) {
    handleEventBridgeError(error);
  }
}

async function createRule(
  eventBridge: EventBridge,
  eventBusName: string,
  notificationType: string,
) {
  try {
    await eventBridge.putRule({
      Name: buildEventBusFilterRuleName(notificationType),
      EventBusName: eventBusName,
      Description: getEventBusRuleDescription(notificationType),
      EventPattern: getSPAPIEventPatternForEventBusRule(notificationType),
      State: RuleState.ENABLED,
    });
  } catch (error: any) {
    if (!(error instanceof ResourceAlreadyExistsException)) {
      handleEventBridgeError(error);
    }
  }
}

/**
 * Helper method which removes the target and deletes the rule.
 * @param eventBusName name of the event bus.
 * @param rule the event bus rule to delete.
 * @param targetId the Id of the target to remove from the rule.
 * @param region the AWS region.
 */
export async function removeTargetAndDeleteRule(
  eventBusName: string,
  rule: string,
  targetId: string,
  region: string,
) {
  const eventBridge = buildEventBridgeClient(region);
  await removeTarget(eventBridge, eventBusName, rule, targetId);
  await deleteRule(eventBridge, eventBusName, rule);
}

async function createAndAssociateEventBusHelper(
  eventBridge: EventBridge,
  eventBusName: string,
) {
  try {
    await eventBridge.createEventBus({
      Name: eventBusName,
      EventSourceName: eventBusName,
    });
  } catch (error: any) {
    if (!(error instanceof ResourceAlreadyExistsException)) {
      handleEventBridgeError(error);
    }
  }
}

/**
 * Helper method to create(If one does not exist already) and associate Event Bus.
 *
 * @param eventBusName EventBus Name.
 * @param region AWS region.
 */
export async function createAndAssociateEventBus(
  eventBusName: string,
  region: string,
) {
  const eventBridge = buildEventBridgeClient(region);
  await createAndAssociateEventBusHelper(eventBridge, eventBusName);
}

/**
 * Helper method to create the necessary Event Bridge resources
 * 1. Create New Event Bus Rule.
 * 2. Create New SQS Queue(For each notification type, if not created already).
 * 3. Create New Target.
 *
 * @param eventBusName EventBus Name.
 * @param notificationType Type of the Notification.
 * @param awsAccountId AWS Account associated with the application.
 * @param region AWS region.
 */
export async function createEventBridgeResources(
  eventBusName: string,
  notificationType: string,
  awsAccountId: string,
  region: string,
) {
  const eventBridge = buildEventBridgeClient(region);
  await createRule(eventBridge, eventBusName, notificationType);
  await createQueue(notificationType, region, awsAccountId, eventBusName);
  await createTarget(
    eventBridge,
    eventBusName,
    region,
    awsAccountId,
    notificationType,
  );
}

function getEventBusRuleDescription(notificationType: string) {
  return `Event Bus rule to receive the ${notificationType} notifications from the selling partner API.`;
}

// Function to get the SP-API Event Pattern for EventBus Rule.
// https://tinyurl.com/bdf98t9f
function getSPAPIEventPatternForEventBusRule(notificationType: string) {
  return `
    {
      "source": [
        {
          "prefix": "aws.partner/sellingpartnerapi.amazon.com"
        }
      ],
      "detail-type": ["${notificationType}"]
    }
  `;
}
