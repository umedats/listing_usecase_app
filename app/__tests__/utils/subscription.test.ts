import getSPAPIEndpoint, {
  buildGrantLessNotificationsAPIClient,
  buildNotificationsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { NotificationsApi } from "@/app/sdk/notifications";
import { MOCK_SUBSCRIPTION_1 } from "@/app/test-utils/mock-subscription";
import {
  createSubscription,
  deleteSubscription,
  getAllListingRawSubscriptions,
  getDisplayableListingSubscriptions,
} from "@/app/utils/subscription";
import { SPAPIRequestResponse } from "@/app/model/types";
import { getEventBridgeDestinations } from "@/app/utils/destination";
import { getRuleTargetsForEventBus } from "@/app/utils/eventbridge";
import {
  buildEventBusFilterRuleName,
  buildNotificationSQSQueueName,
  buildNotificationSQSTargetName,
} from "@/app/utils/aws";
import {
  IPTC_NOTIFICATION_TYPE,
  LIIC_NOTIFICATION_TYPE,
} from "@/app/constants/global";

const SUBSCRIPTION_ID = "0c49f1df-6dab-4ac1-a887-87db417952d3";
const EVENT_BUS_NAME =
  "sellingpartnerapi.amazon.com/amzn1.sellerapps.app.15a75829-cd4a-4efc-b947-0cc39d874577";
const AWS_ACCOUNT_ID = "123456789";
const DESTINATION_ID = "0c43f1df-6dab-4ac1-a887-87db417972d3";

jest.mock("@/app/sdk/factory/ListingsSPAPIClientsFactory");
const mockedGetSPAPIEndpoint = jest.mocked(getSPAPIEndpoint);
mockedGetSPAPIEndpoint.mockReturnValue("");

const notificationsApi = new NotificationsApi(undefined);
const grantlessNotificationsApi = new NotificationsApi(undefined);
const mockedGetSubscription = jest.fn();
const mockedDeleteSubscriptionById = jest.fn();
const mockedCreateSubscription = jest.fn();
notificationsApi.getSubscription = mockedGetSubscription;
notificationsApi.createSubscription = mockedCreateSubscription;
grantlessNotificationsApi.deleteSubscriptionById = mockedDeleteSubscriptionById;

const mockedBuildNotificationsAPIClient = jest.mocked(
  buildNotificationsAPIClient,
);
const mockedBuildGrantLessNotificationsAPIClient = jest.mocked(
  buildGrantLessNotificationsAPIClient,
);
mockedBuildNotificationsAPIClient.mockResolvedValue(notificationsApi);
mockedBuildGrantLessNotificationsAPIClient.mockResolvedValue(
  grantlessNotificationsApi,
);

jest.mock("@/app/utils/destination");
const mockedGetEventBridgeDestinations = jest.mocked(
  getEventBridgeDestinations,
);

jest.mock("@/app/utils/eventbridge");
const mockedGetRuleTargetsForEventBus = jest.mocked(getRuleTargetsForEventBus);

const REGION = "us-east-1";
describe("Test for the subscription", () => {
  describe("Test for the deleteSubscription", () => {
    test("snapshot test for 200 response from DeleteSubscriptionById API", async () => {
      mockedDeleteSubscriptionById.mockResolvedValue({});

      const reqResponses: SPAPIRequestResponse[] = [];
      await deleteSubscription(
        SUBSCRIPTION_ID,
        IPTC_NOTIFICATION_TYPE,
        REGION,
        reqResponses,
      );
      expect(reqResponses).toMatchSnapshot();
    });

    test("snapshot test for 404 response from DeleteSubscriptionById API", async () => {
      mockedDeleteSubscriptionById.mockRejectedValue({
        status: 404,
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      await deleteSubscription(
        SUBSCRIPTION_ID,
        IPTC_NOTIFICATION_TYPE,
        REGION,
        reqResponses,
      );
      expect(reqResponses).toMatchSnapshot();
    });

    test("snapshot test for 500 response from DeleteSubscriptionById API", async () => {
      mockedDeleteSubscriptionById.mockRejectedValue({
        status: 500,
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      await expect(
        deleteSubscription(
          SUBSCRIPTION_ID,
          IPTC_NOTIFICATION_TYPE,
          REGION,
          reqResponses,
        ),
      ).rejects.toMatchObject({
        status: 500,
      });
      expect(reqResponses).toMatchSnapshot();
    });
  });

  describe("Test for the getAllListingRawSubscriptions", () => {
    test("snapshot test for 200 and 404 responses from GetSubscription API", async () => {
      mockedGetSubscription
        .mockResolvedValueOnce({
          payload: {
            subscriptionId: MOCK_SUBSCRIPTION_1.subscriptionId,
            payloadVersion: "1.0",
            destinationId: MOCK_SUBSCRIPTION_1.subscriptionId,
          },
        })
        .mockRejectedValue({
          status: 404,
        });

      const reqResponses: SPAPIRequestResponse[] = [];
      const rawSubscriptions = await getAllListingRawSubscriptions(
        REGION,
        false,
        reqResponses,
      );
      expect(rawSubscriptions).toMatchSnapshot();
      expect(reqResponses).toMatchSnapshot();
    });

    test("snapshot test for the 500 response from GetSubscription API", async () => {
      mockedGetSubscription.mockRejectedValue({
        status: 500,
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      await expect(
        getAllListingRawSubscriptions(REGION, false, reqResponses),
      ).rejects.toMatchObject({
        status: 500,
      });
      expect(reqResponses).toMatchSnapshot();
    });
  });

  function mockGetSubscriptionForDisplayableSubscriptionsTest() {
    mockedGetSubscription
      // UnMatched
      .mockResolvedValueOnce({
        payload: {
          subscriptionId: "99821be8-8dfe-4c1c-a1f7-895ab790cb4f",
          payloadVersion: "1.0",
          destinationId: "40807127-05cf-4e75-b833-1778eb7d9530",
        },
      })
      // Matched with two rules
      .mockResolvedValueOnce({
        payload: {
          subscriptionId: "5f21f928-ca44-414c-a5c6-6a3729cb6b5c",
          payloadVersion: "1.0",
          destinationId: "f0195c18-3ff4-471e-8618-b406e0f37cbc",
        },
      })
      // Matched with no rules
      .mockResolvedValueOnce({
        payload: {
          subscriptionId: "83d66cc2-9063-43f7-b1a1-391a7aa0bba8",
          payloadVersion: "1.0",
          destinationId: "f0195c18-3ff4-471e-8618-b406e0f37cbc",
        },
      })
      // Matched with one rule
      .mockResolvedValueOnce({
        payload: {
          subscriptionId: "04fb6be6-c49a-4016-be54-2da408b91175",
          payloadVersion: "1.0",
          destinationId: "f0195c18-3ff4-471e-8618-b406e0f37cbc",
        },
      })
      // UnMatched
      .mockResolvedValueOnce({
        payload: {
          subscriptionId: "b541bb7f-4581-43d0-ab6b-c2ad63cd3bfe",
          payloadVersion: "1.0",
          destinationId: "72286d85-66b3-47e2-9762-88baaf9aef05",
        },
      })
      .mockRejectedValue({
        status: 404,
      });
  }

  describe("Test for the getDisplayableListingSubscriptions", () => {
    test("snapshot test which covers all the major edge cases", async () => {
      mockGetSubscriptionForDisplayableSubscriptionsTest();

      mockedGetEventBridgeDestinations.mockResolvedValue([
        {
          destinationId: "f0195c18-3ff4-471e-8618-b406e0f37cbc",
          name: "TestEventBridgeDestination",
          resource: {
            eventBridge: {
              name: EVENT_BUS_NAME,
              accountId: AWS_ACCOUNT_ID,
              region: REGION,
            },
          },
        },
      ]);

      mockedGetRuleTargetsForEventBus.mockResolvedValue([
        {
          rule: {
            Name: buildEventBusFilterRuleName(IPTC_NOTIFICATION_TYPE),
            EventBusName: EVENT_BUS_NAME,
          },
          target: {
            Id: buildNotificationSQSTargetName(IPTC_NOTIFICATION_TYPE),
            Arn: `arn:aws:sqs:${REGION}:${AWS_ACCOUNT_ID}:${buildNotificationSQSQueueName(
              IPTC_NOTIFICATION_TYPE,
            )}`,
          },
        },
        {
          rule: {
            Name: buildEventBusFilterRuleName(IPTC_NOTIFICATION_TYPE),
            EventBusName: EVENT_BUS_NAME,
          },
          target: {
            Id: "ItemProductTypeChangeTarget1",
            Arn: `arn:aws:sqs:${REGION}:${AWS_ACCOUNT_ID}:${buildNotificationSQSQueueName(
              IPTC_NOTIFICATION_TYPE,
            )}`,
          },
        },
        {
          rule: {
            Name: buildEventBusFilterRuleName(LIIC_NOTIFICATION_TYPE),
            EventBusName: EVENT_BUS_NAME,
          },
          target: {
            Id: buildNotificationSQSTargetName(LIIC_NOTIFICATION_TYPE),
            Arn: `arn:aws:sqs:${REGION}:${AWS_ACCOUNT_ID}:${buildNotificationSQSQueueName(
              LIIC_NOTIFICATION_TYPE,
            )}`,
          },
        },
      ]);

      const reqResponses: any[] = [];
      const displayableSubscriptions = await getDisplayableListingSubscriptions(
        AWS_ACCOUNT_ID,
        REGION,
        true,
        reqResponses,
      );
      expect(displayableSubscriptions).toMatchSnapshot();
    });

    test("snapshot test for no destinations", async () => {
      mockGetSubscriptionForDisplayableSubscriptionsTest();

      mockedGetEventBridgeDestinations.mockResolvedValue([]);

      const reqResponses: any[] = [];
      const displayableSubscriptions = await getDisplayableListingSubscriptions(
        AWS_ACCOUNT_ID,
        REGION,
        true,
        reqResponses,
      );
      expect(displayableSubscriptions).toMatchSnapshot();
    });
  });

  describe("Test for the create Subscription", () => {
    test("snapshot test for 200 response from CreateSubscription API", async () => {
      mockedCreateSubscription.mockResolvedValue({});

      const reqResponses: SPAPIRequestResponse[] = [];
      await createSubscription(
        REGION,
        DESTINATION_ID,
        IPTC_NOTIFICATION_TYPE,
        reqResponses,
      );

      expect(reqResponses).toMatchSnapshot();
    });

    test("snapshot test for 409 Conflict response from CreateSubscription API", async () => {
      mockedCreateSubscription.mockRejectedValue({
        status: 409,
      });

      const reqResponses: SPAPIRequestResponse[] = [];
      await expect(
        createSubscription(
          REGION,
          DESTINATION_ID,
          IPTC_NOTIFICATION_TYPE,
          reqResponses,
        ),
      ).rejects.toMatchObject({
        status: 409,
      });

      expect(reqResponses).toMatchSnapshot();
    });
  });
});
