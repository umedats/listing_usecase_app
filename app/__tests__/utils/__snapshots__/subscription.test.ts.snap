// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the subscription Test for the create Subscription snapshot test for 200 response from CreateSubscription API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#createsubscription",
    "apiName": "CreateSubscription",
    "request": {
      "body": CreateSubscriptionRequest {
        "destinationId": "0c43f1df-6dab-4ac1-a887-87db417972d3",
        "payloadVersion": "1.0",
      },
      "notificationType": "ITEM_PRODUCT_TYPE_CHANGE",
    },
    "response": {},
  },
]
`;

exports[`Test for the subscription Test for the create Subscription snapshot test for 409 Conflict response from CreateSubscription API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#createsubscription",
    "apiName": "CreateSubscription",
    "request": {
      "body": CreateSubscriptionRequest {
        "destinationId": "0c43f1df-6dab-4ac1-a887-87db417972d3",
        "payloadVersion": "1.0",
      },
      "notificationType": "ITEM_PRODUCT_TYPE_CHANGE",
    },
    "response": {
      "status": 409,
    },
  },
]
`;

exports[`Test for the subscription Test for the deleteSubscription snapshot test for 200 response from DeleteSubscriptionById API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#deletesubscriptionbyid",
    "apiName": "DeleteSubscriptionById",
    "request": {
      "notificationType": "ITEM_PRODUCT_TYPE_CHANGE",
      "subscriptionId": "0c49f1df-6dab-4ac1-a887-87db417952d3",
    },
    "response": {},
  },
]
`;

exports[`Test for the subscription Test for the deleteSubscription snapshot test for 404 response from DeleteSubscriptionById API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#deletesubscriptionbyid",
    "apiName": "DeleteSubscriptionById",
    "request": {
      "notificationType": "ITEM_PRODUCT_TYPE_CHANGE",
      "subscriptionId": "0c49f1df-6dab-4ac1-a887-87db417952d3",
    },
    "response": {
      "status": 404,
    },
  },
]
`;

exports[`Test for the subscription Test for the deleteSubscription snapshot test for 500 response from DeleteSubscriptionById API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#deletesubscriptionbyid",
    "apiName": "DeleteSubscriptionById",
    "request": {
      "notificationType": "ITEM_PRODUCT_TYPE_CHANGE",
      "subscriptionId": "0c49f1df-6dab-4ac1-a887-87db417952d3",
    },
    "response": {
      "status": 500,
    },
  },
]
`;

exports[`Test for the subscription Test for the getAllListingRawSubscriptions snapshot test for 200 and 404 responses from GetSubscription API 1`] = `
[
  {
    "notificationType": "LISTINGS_ITEM_ISSUES_CHANGE",
    "subscription": Subscription {
      "destinationId": "f8c3de3d-1fea-4d7c-a8b0-29f63c4c3454",
      "payloadVersion": "1.0",
      "subscriptionId": "f8c3de3d-1fea-4d7c-a8b0-29f63c4c3454",
    },
  },
]
`;

exports[`Test for the subscription Test for the getAllListingRawSubscriptions snapshot test for 200 and 404 responses from GetSubscription API 2`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#getsubscription",
    "apiName": "GetSubscription",
    "request": {
      "notificationType": "LISTINGS_ITEM_ISSUES_CHANGE",
    },
    "response": GetSubscriptionResponse {
      "payload": Subscription {
        "destinationId": "f8c3de3d-1fea-4d7c-a8b0-29f63c4c3454",
        "payloadVersion": "1.0",
        "subscriptionId": "f8c3de3d-1fea-4d7c-a8b0-29f63c4c3454",
      },
    },
  },
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#getsubscription",
    "apiName": "GetSubscription",
    "request": {
      "notificationType": "PRODUCT_TYPE_DEFINITIONS_CHANGE",
    },
    "response": {
      "status": 404,
    },
  },
]
`;

exports[`Test for the subscription Test for the getAllListingRawSubscriptions snapshot test for the 500 response from GetSubscription API 1`] = `
[
  {
    "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/notifications-api-v1-reference#getsubscription",
    "apiName": "GetSubscription",
    "request": {
      "notificationType": "LISTINGS_ITEM_ISSUES_CHANGE",
    },
    "response": {
      "status": 500,
    },
  },
]
`;

exports[`Test for the subscription Test for the getDisplayableListingSubscriptions snapshot test for no destinations 1`] = `
[
  {
    "notificationType": "BRANDED_ITEM_CONTENT_CHANGE",
    "subscriptionId": "99821be8-8dfe-4c1c-a1f7-895ab790cb4f",
  },
  {
    "notificationType": "ITEM_PRODUCT_TYPE_CHANGE",
    "subscriptionId": "5f21f928-ca44-414c-a5c6-6a3729cb6b5c",
  },
  {
    "notificationType": "LISTINGS_ITEM_STATUS_CHANGE",
    "subscriptionId": "83d66cc2-9063-43f7-b1a1-391a7aa0bba8",
  },
  {
    "notificationType": "LISTINGS_ITEM_ISSUES_CHANGE",
    "subscriptionId": "04fb6be6-c49a-4016-be54-2da408b91175",
  },
  {
    "notificationType": "LISTINGS_ITEM_MFN_QUANTITY_CHANGE",
    "subscriptionId": "b541bb7f-4581-43d0-ab6b-c2ad63cd3bfe",
  },
]
`;

exports[`Test for the subscription Test for the getDisplayableListingSubscriptions snapshot test which covers all the major edge cases 1`] = `
[
  {
    "eventBusName": "sellingpartnerapi.amazon.com/amzn1.sellerapps.app.15a75829-cd4a-4efc-b947-0cc39d874577",
    "notificationType": "ITEM_PRODUCT_TYPE_CHANGE",
    "subscriptionId": "5f21f928-ca44-414c-a5c6-6a3729cb6b5c",
  },
  {
    "eventBusName": "sellingpartnerapi.amazon.com/amzn1.sellerapps.app.15a75829-cd4a-4efc-b947-0cc39d874577",
    "notificationType": "LISTINGS_ITEM_STATUS_CHANGE",
    "subscriptionId": "83d66cc2-9063-43f7-b1a1-391a7aa0bba8",
  },
  {
    "eventBusFilterRule": "ListingsItemIssuesChangeRule",
    "eventBusName": "sellingpartnerapi.amazon.com/amzn1.sellerapps.app.15a75829-cd4a-4efc-b947-0cc39d874577",
    "notificationType": "LISTINGS_ITEM_ISSUES_CHANGE",
    "sqsQueueName": "ListingsItemIssuesChangeQueue",
    "sqsTargetId": "ListingsItemIssuesChangeTarget",
    "subscriptionId": "04fb6be6-c49a-4016-be54-2da408b91175",
  },
  {
    "notificationType": "BRANDED_ITEM_CONTENT_CHANGE",
    "subscriptionId": "99821be8-8dfe-4c1c-a1f7-895ab790cb4f",
  },
  {
    "notificationType": "LISTINGS_ITEM_MFN_QUANTITY_CHANGE",
    "subscriptionId": "b541bb7f-4581-43d0-ab6b-c2ad63cd3bfe",
  },
]
`;
